Add "Access instance" section to "Work with Android" page (#104)
canonical/anbox-cloud-docs: howto/android/access-instance.md

(howto-access-instance)=
# How to access an instance

For certain scenarios, accessing an individual instance for debugging or development might be necessary. You can do this by:

- Running `anbox-connect` for secure Android Debug Bridge (ADB) connections.

> - Exposing the ADB service upon Anbox instance creation.

Reviewer: ```suggestion
- Exposing the ADB service upon Anbox instance creation. We use the [`scrcpy`](https://github.com/Genymobile/scrcpy) tool for this method.
```

Reviewer: Okay, I understood your point from the SU discussion. We can keep the classification to Not exposing the service (Anbox connect) vs Exposing ADB service.

```{note}
Each presigned URL can only be used to establish a single ADB connection. If multiple users attempt to use the same presigned URL, any existing ADB connection will be interrupted to allow the new request to succeed.
```

### Connect to the Instance

Open a terminal and use the `anbox-connect` command to establish a secure ADB connection to the remote Android instance:

    anbox-connect <connection_url>

After the ADB channel is established, the following output will be displayed:

    Access to the remote Anbox Cloud instance over ADB is now possible. Please run

        $ adb connect 127.0.0.1:32985

Lastly, follow the prompt in the command line output and run the following command to set up the ADB connection:

    adb connect 127.0.0.1:32985

```{note}
The `anbox-connect` command sets up a secure ADB channel and routes traffic between your local machine and the remote Anbox instance. Therefore, it must be kept running to maintain the ADB connection. Do not abort the command once the connection is established.
```

(sec-expose-adb-service)=
## Expose ADB service upon Anbox instance creation

> The AMS services allow for remote connections to instances over a network by exposing the [ADB service](https://anbox-cloud.io/docs/howto/instance/expose-services).

Reviewer: ```suggestion
The AMS services enable connecting remotely over a network to instance using scrcpy. Scrcpy provides a simple way to display and control instances running on Anbox Cloud remotely.
```

(howto-access-instance)=
# How to access an instance

For certain scenarios, accessing an individual instance for debugging or development might be necessary. You can do this by:

- Run `anbox-connect` for secure ADB connections.
- Expose ADB service upon Anbox instance creation.

## Access the Android instance using anbox-connect

Since the 1.23.0 release, Anbox Cloud allows users to create a secure ADB connection through the `anbox-connect` command without exposing the [ADB service](https://anbox-cloud.io/docs/howto/instance/expose-services). This method is highly recommended for securely connecting to remote Android instances, offering a safe and efficient means of managing ADB connections.

### Install `anbox-connect` snap

First, install the `anbox-connect` snap from the snap store:

    snap install anbox-connect

### Generate a presigned connection URL

Generate a presigned connection URL in the following ways:

- **From the Dashboard**:

>   Navigate to the `Instances` view from the left panel, locate a running instance, click the ![Connect ADB|16x16](https://assets.ubuntu.com/v1/51973cea-adb-connect-icon.png) icon button, then click `Authorize` in the popup dialog and copy the command provided.

Reviewer: Just to be consistent with how we use inlines icons:
```suggestion
In the *Instances* list view, locate a running instance and click *Connect ADB* ( ![Connect ADB|16x16](https://assets.ubuntu.com/v1/51973cea-adb-connect-icon.png) ).
*Authorize* the connection and copy the command provided.
```

- **From the Command Line**:
  * For a full Anbox Cloud deployment:

        anbox-stream-gateway session share <session_id>

  * For the Anbox Cloud Appliance:

        anbox-cloud-appliance.gateway session share <session_id>

  This command will return a presigned URL that you can use to connect to the remote Android instance.

```{note}
Each presigned URL can only be used to establish a single ADB connection. If multiple users attempt to use the same presigned URL, any existing ADB connection will be interrupted to allow the new request to succeed.
```

### Connect to the Instance

Open a terminal and use the `anbox-connect` command to establish a secure ADB connection to the remote Android instance:

    anbox-connect <connection_url>

```{note}
If the Anbox Stream Gateway or Anbox Cloud Appliance uses self-signed certificates, include the `--skip-tls-verify` or `-k` parameter to skip TLS certificate verification.
```

After the ADB channel is established, the following output will be displayed:

    Access to the remote Anbox Cloud instance over ADB is now possible. Please run

        $ adb connect 127.0.0.1:32985

> Lastly, follow the prompt in the command line output and run:
> 
>     adb connect 127.0.0.1:32985
> 
> to setup the ADB connection.

Reviewer: To make it look continuous in the output:
```suggestion
Lastly, follow the prompt in the command line output and run the following command to set up the ADB connection:

    adb connect 127.0.0.1:32985
```

```{note}
The `anbox-connect` command sets up a secure ADB channel and routes traffic between your local machine and the remote Anbox instance. Therefore, it must be kept running to maintain the ADB connection. Do not abort the command once the connection is established.
```

(sec-expose-adb-service)=

> ## Expose ADB service upon Anbox instance creation

Reviewer: Two suggestions:
1. It would be a good idea to call out that we use a third party tool, scrcpy: 
Something similar to what was existing - "The AMS services enable connecting remotely over a network to instance via [scrcpy](https://github.com/Genymobile/scrcpy).Scrcpy provides a simple way to display and control instances running on Anbox Cloud remotely."

I understand users will realise when they are performing the task but it is better to let them know before they start the task.
2. Although this method is not preferred as it not secure, I think we should mention that this method provides graphical access

Reviewer: @adglkh I have added suggestions to implement what I meant in this comment. Please have a look, everything else is good.

### Connect to the Instance

Connect the running instance remotely, establish the connection between your host and the instance through the exposed ADB service via TCP/IP:

    adb connect 10.226.4.200:10000

After the connection is established, you can dump the Android system logs with:

    adb logcat

Or simply run [scrcpy](https://github.com/Genymobile/scrcpy) for the display mirror, which will return information similar to the following:

```bash
INFO: scrcpy 1.10 <https://github.com/Genymobile/scrcpy>
/usr/local/share/scrcpy/scrcpy-server.jar: 1 file pushed. 9.3 MB/s (22662 bytes in 0.002s)
INFO: Initial texture: 1280x720
```

Then you can interact with the running Android container locally.

### Through SSH tunnel

In the above example, the ADB service is exposed directly over the internet. This is a major security risk as the ADB connection is not secure. To overcome this security issue, you can use the machine where AMS is running as relay server to set up a secure and encrypted SSH tunnel by forwarding the exposed ADB TCP port from the LXD machine to your localhost through the AMS machine.

```{note}

> Creating an ADB connection through an SSH tunnel is not straightforward and can be troublesome. It's strongly recommanded to run anbox-connect to setup a secure ADB connection for the remote access to Android instance.

Reviewer: ```suggestion
Creating an ADB connection through an SSH tunnel is not straightforward and can be troublesome. It's strongly recommended to run anbox-connect to setup a secure ADB connection for the remote access to Android instance.
```
